src/ship.h

```diff
@@ -12,17 +12,23 @@
 #ifndef SHIP_H
 #define SHIP_H
 
+#include <deque>
+
 #include "vehicle_base.h"
 #include "water_map.h"
 
 void GetShipSpriteSize(EngineID engine, uint &width, uint &height, int &xoffs, int &yoffs, EngineImageType image_type);
 WaterClass GetEffectiveWaterClass(TileIndex tile);
 
+typedef std::deque<TrackdirByte> ShipPathCache;
+
 /**
  * All ships have this type.
  */
 struct Ship FINAL : public SpecializedVehicle<Ship, VEH_SHIP> {
 	TrackBitsByte state; ///< The "track" the ship is following.
+	ShipPathCache path;  ///< Cached path.
+	TileIndex path_dest; ///< Cached path destination.
```

Comment:
Hmm, what about virtual methods? This [commit](https://github.com/PeterN/OpenTTD/commit/f17d37f0a11adf03568e014f8a3d52aac30ffe48) implements this.

Reply:
Yes, it will call the virtual function on Ship as expected. There is a small performance penalty for this but I don't think that's relevant here.

Reply:
Just to be sure...
If I call UpdateOrderDest(Vehicle *v, const Order *order, int conditional_depth,...
and inside UpdateOrderDest(...) there is a call to:
v->SetDest()
It will call the virtual method Ship::SetDest and not Vehicle::SetDest ?

Reply:
Yup.

Reply:
Then, after all this hassle, is up to you to decide whether it is best to use path_dest or clearing cache when dest_tile changes.
I haven't reviewed the code for load/saving dequeues, but the other part of the code looks great.
This is a great improvement and I hope you have it soon approved.